e2e/backend_service_update_test.go

```diff
@@ -226,3 +255,86 @@ func TestBackendServiceUpdate(t *testing.T) {
 	wantBS = &compute.BackendService{LoadBalancingScheme: "INTERNAL_MANAGED"}
 	checkBackendService(t, ctx, theCloud, bs1ID, wantBS, compareLBScheme)
 }
+
+func TestBackendServiceMetadataUpdate(t *testing.T) {
+	t.Parallel()
+
+	ctx := context.Background()
+	graphBuilder := rgraph.NewBuilder()
+	resUniqueIdPart := "test-bs-metadata-update"
+
+	hcName := resourceName(resUniqueIdPart + "-hc")
+	hcID, err := buildHealthCheck(graphBuilder, hcName, 15)
+	if err != nil {
+		t.Fatalf("buildHealthCheck(_, %s, _) = (_, %v) want (_, nil)", hcName, err)
+	}
+
+	bsName := resourceName(resUniqueIdPart + "-bs")
+	bsID, err := buildBackendServiceWithMetadata(graphBuilder, bsName, hcID, map[string]string{"foo": "1", "bar": "2", "baz": "2"})
+
+	expectedActions := []exec.ActionMetadata{
+		{Type: exec.ActionTypeCreate, Name: actionName(exec.ActionTypeCreate, hcID)},
+		{Type: exec.ActionTypeCreate, Name: actionName(exec.ActionTypeCreate, bsID)},
+	}
+
+	processGraphAndExpectActions(t, graphBuilder, expectedActions)
+	t.Cleanup(func() {
+		err = theCloud.BackendServices().Delete(ctx, bsID.Key)
+		if err != nil {
+			t.Logf("theCloud.BackendServices().Delete(_, %s): %v", bsID.Key, err)
+		}
+		err = theCloud.HealthChecks().Delete(ctx, hcID.Key)
+		if err != nil {
+			t.Logf("theCloud.HealthChecks().Delete(_, %s): %v", hcID.Key, err)
+		}
+	})
+	_, err = theCloud.HealthChecks().Get(ctx, hcID.Key)
+	if err != nil {
+		t.Fatalf("theCloud.HealthChecks().Get(_, %s) = (nil, %v) wanted (_, nil)", hcID.Key, err)
+	}
+	bs, err := theCloud.BackendServices().Get(ctx, bsID.Key)
+	if err != nil {
+		t.Fatalf("theCloud.BackendServices().Get(_, %s) = (nil, %v) wanted (_, nil)", bsID.Key, err)
+	}
+
+	wantedMetadata := map[string]string{"foo": "1", "bar": "2", "baz": "2"}
+	if df := cmp.Diff(bs.Metadatas, wantedMetadata); df != "" {
+		t.Fatalf("Metadata are not set correctly: %s", df)
+	}
+
+	bsID, err = buildBackendServiceWithMetadata(graphBuilder, bsName, hcID, map[string]string{"foo": "1", "bar": "2", "baz": "3"})
+	expectedActions = []exec.ActionMetadata{
+		{Type: exec.ActionTypeMeta, Name: eventName(hcID)},
+		{Type: exec.ActionTypeUpdate, Name: actionName(exec.ActionTypeUpdate, bsID)},
+	}
+
+	processGraphAndExpectActions(t, graphBuilder, expectedActions)
+	wantedMetadata = map[string]string{"foo": "1", "bar": "2", "baz": "3"}
+
+	bs, err = theCloud.BackendServices().Get(ctx, bsID.Key)
+	if err != nil {
+		t.Fatalf("theCloud.BackendServices().Get(_, %s) = (nil, %v) wanted (_, nil)", bsID.Key, err)
+	}
+
+	if df := cmp.Diff(bs.Metadatas, wantedMetadata); df != "" {
+		t.Fatalf("Metadata keys are not updated correctly(-got, +want): %s", df)
+	}
+
+	bsID, err = buildBackendServiceWithMetadata(graphBuilder, bsName, hcID, map[string]string{"foo": "1", "bar": "2"})
```

Comment:
pleas add comments to each update operation what is this updating like: delete key, update one key etc

Reply:
Done.